hcl2template/types.packer_config.go

```diff
@@ -280,13 +285,89 @@ func (cfg *PackerConfig) getCoreBuildPostProcessors(source SourceBlock, blocks [
 	return res, diags
 }
 
+type provisionerPrepare struct {
+	cfg               *PackerConfig
+	provisionersBlock []*ProvisionerBlock
+	src               SourceRef
+}
+
+// HCL2ProvisionerPrepare is used by the ProvisionHook at the runtime in the provision step
+// to interpolate any build variable by decoding and preparing it again.
+func (pp *provisionerPrepare) HCL2ProvisionerPrepare(data map[string]interface{}) ([]packer.CoreBuildProvisioner, hcl.Diagnostics) {
+	src := pp.cfg.Sources[pp.src.Ref()]
+
+	variables := make(Variables)
+	for k, v := range data {
+		if value, ok := v.(string); ok {
+			variables[k] = &Variable{
+				DefaultValue: cty.StringVal(value),
+				Type:         cty.String,
+			}
+		}
+	}
+	variablesVal, _ := variables.Values()
+
+	generatedVariables := map[string]cty.Value{
+		sourcesAccessor: cty.ObjectVal(map[string]cty.Value{
+			"type": cty.StringVal(src.Type),
+			"name": cty.StringVal(src.Name),
+		}),
+		buildAccessor: cty.ObjectVal(variablesVal),
+	}
+
+	return pp.cfg.getCoreBuildProvisioners(src, pp.provisionersBlock, pp.cfg.EvalContext(generatedVariables))
+}
+
+type postProcessorsPrepare struct {
```

Comment:
do you think this and `provisionerPrepare` could be in another file, ( I'm feeling like this file is getting huuuge )

Reply:
I can move them to `types.build.provisioners` and `types.build.post-processor` 🤔  I think it makes sense!